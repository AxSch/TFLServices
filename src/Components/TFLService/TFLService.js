import React, { useState, useEffect } from 'react'

const TFLService = ({ service }) => {
    const [isDisrupted, setIsDisrupted] = useState(false)
    const severeLineStatuses = service.lineStatuses.filter(line => line.statusSeverity !== 10)

    useEffect(() => {
        if (severeLineStatuses.length > 0) {
            if (severeLineStatuses[0].statusSeverity !== 10 ) {
                setIsDisrupted(true)
            }
        } else {
            setIsDisrupted(false)
        }
    }, [isDisrupted, severeLineStatuses])

    const renderDisruptionHeader = () => {
        if (severeLineStatuses.length > 0) {
            return (
                <span>
                    Service is currently suffering disruptions
                </span>
            )
        } else {
            return (
                <span>
                    No service disruptions
                </span>
            )
        }
    }

    const renderDisruptionsList = () => {
        if (severeLineStatuses.length > 0) {
            return severeLineStatuses.map((line, index) => {
                return (
                    <li key={index}>
                        {line.reason}
                    </li>
                )
            })
        }
    }
    const renderDisruptions = () => {
        if (isDisrupted) {
            return (
                <div>
                    <h3>Disruptions:</h3>
                    <ul>
                        {renderDisruptionsList()}
                    </ul>
                </div>
            )
        }
    }
    // console.log(isDisrupted)
    return (
        <div>
            <div>
                <h2>{service.name}</h2>
                <span>{renderDisruptionHeader()}</span>
            </div>
            {renderDisruptions()}
        </div>
    )
}

export default TFLService
