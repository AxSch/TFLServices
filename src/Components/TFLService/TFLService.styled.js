import styled from 'styled-components'

const TFLServiceContainer = styled.div`
    margin-top: 1.2rem;
    display: flex;
    justify-content: center;
    width: 100%;
    flex-wrap: wrap;
    padding: 1rem;

    @media(min-width: 768px) {
        margin-top: 0;
        align-content: start;
    }
    
    @media(min-width: 1024px) {
        justify-content: start;
        width: 50%;
    }
`
const ServiceHeader = styled.div`
    display: flex;
    flex-direction: column;

    @media(min-width: 1024px) {
        align-items: start;
    }
`

const ServiceDisruptions = styled.div`
    display: flex;
    flex-wrap: wrap;
    justify-content: center;

    @media(min-width: 1024px) {
        align-content: start;
        justify-content: start;
    }
`

const ServiceDisruptionsList = styled.ul`
    list-style-type: none;
    text-align: left;
    padding-left: 1.4rem;

    @media(min-width: 1024px) {
        padding-left: 0; 
    }
`

const DisruptionsListItem = styled.li`
    margin: 1.4rem 0;
`


export {
    ServiceHeader,
    TFLServiceContainer,
    ServiceDisruptions,
    ServiceDisruptionsList,
    DisruptionsListItem,
}
