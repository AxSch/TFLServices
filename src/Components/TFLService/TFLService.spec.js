import React from 'react'
import { render } from '@testing-library/react';
import TFLService from './TFLService'

const mockService = {
    $type: "Tfl.Api.Presentation.Entities.Line, Tfl.Api.Presentation.Entities",
    created: "2021-06-03T11:02:50.87Z",
    crowding: {$type: "Tfl.Api.Presentation.Entities.Crowding, Tfl.Api.Presentation.Entities"},
    disruptions: [],
    id: "london-overground",
    lineStatuses: [
        {
            $type: "Tfl.Api.Presentation.Entities.LineStatus, Tfl.Api.Presentation.Entities",
            created: "0001-01-01T00:00:00",
            disruption: {$type: "Tfl.Api.Presentation.Entities.Disruption, Tfl.Api.Presentation.Entities", category: "PlannedWork", categoryDescription: "PlannedWork", description: "Changes to Stratford to Richmond services on Mondaâ€¦5:56, 17:08, 18:20, 19:32, 20:47, 22:01 and 23:17", created: "2021-05-07T08:46:00Z"},
            id: 0,
            lineId: "london-overground",
            reason: "Changes to Stratford to Richmond services on Mondays to Fridays from Monday 17 May until Friday 18 June: The following trains from Stratford to Richmond will not run: 06:05, 07:18, 08:30, 09:40, 10:59, 12:14, 13:29, 14:43, 15:52, 17:05, 18:16, 19:29, 20:44, 21:59 and 23:14. The 22:59 is now the last train from Stratford for stations from Willesden Junction to Richmond and Clapham Junction - later trains will run from Stratford as far as Willesden Junction only. The following trains from Richmond to Stratford will not run: 06:08. 07:21, 08:34, 09:47, 11:01, 12:17, 13:31, 14:45, 15:56, 17:08, 18:20, 19:32, 20:47, 22:01 and 23:17",
            statusSeverity: 7,
            statusSeverityDescription: "Reduced Service",
            validityPeriods: [{}]
        }
    ],
    modeName: "overground",
    modified: "2021-06-03T11:02:50.87Z",
    name: "London Overground",
    routeSections: [{}, {}, {}, {}],
    serviceTypes: [
        {
            $type: "Tfl.Api.Presentation.Entities.LineServiceTypeInfo, Tfl.Api.Presentation.Entities",
            name: "Regular",
            uri: "/Line/Route?ids=London Overground&serviceTypes=Regular"
        },
        {
            $type: "Tfl.Api.Presentation.Entities.LineServiceTypeInfo, Tfl.Api.Presentation.Entities",
            name: "Night",
            uri: "/Line/Route?ids=London Overground&serviceTypes=Night"
        }
    ]
}

describe('TFLService', () => {
    let container, getByTestId

    it('should render successfully', () => {
        const utils = render(<TFLService service={mockService} />);
        container = utils.container;
        getByTestId = utils.getByTestId;
        expect(getByTestId("tflService")).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('should render the list of disruptions', () => {
        const utils = render(<TFLService service={mockService} />);
        container = utils.container;
        getByTestId = utils.getByTestId;
        expect(getByTestId("disruptionsList")).toBeInTheDocument();
        expect(getByTestId("disruptionsList").children).toHaveLength(1);
    });
    
    it('should render correct disruption copy - disruptions exist', () => {
        const utils = render(<TFLService service={mockService} />);
        container = utils.container;
        getByTestId = utils.getByTestId;
        expect(getByTestId("hasDisruptions")).toBeInTheDocument();
        expect(getByTestId("hasDisruptions").textContent).toEqual('Service is currently suffering disruptions');
    });
    
    
    it('should render correct disruption copy - disruptions do not exist', () => {
        mockService.lineStatuses = []
        const utils = render(<TFLService service={mockService} />);
        container = utils.container;
        getByTestId = utils.getByTestId;
        expect(getByTestId("noDisruptions")).toBeInTheDocument();
        expect(getByTestId("noDisruptions").textContent).toEqual('No service disruptions');
    });

})
