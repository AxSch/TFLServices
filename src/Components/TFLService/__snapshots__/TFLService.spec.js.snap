// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TFLService should render successfully 1`] = `
<div>
  <div
    class="sc-bdnxRM bkQdMX"
    data-testid="tflService"
  >
    <div
      class="sc-gtsrHT fHwCUC"
    >
      <h2>
        London Overground
      </h2>
      <span
        data-testid="hasDisruptions"
      >
        Service is currently suffering disruptions
      </span>
    </div>
    <div
      class="sc-dlnjwi cXkqvs"
    >
      <h3>
        Disruptions:
      </h3>
      <ul
        class="sc-hKFxyN jUKCvu"
        data-testid="disruptionsList"
      >
        <li
          class="sc-eCApnc isxqBE"
        >
          Changes to Stratford to Richmond services on Mondays to Fridays from Monday 17 May until Friday 18 June: The following trains from Stratford to Richmond will not run: 06:05, 07:18, 08:30, 09:40, 10:59, 12:14, 13:29, 14:43, 15:52, 17:05, 18:16, 19:29, 20:44, 21:59 and 23:14. The 22:59 is now the last train from Stratford for stations from Willesden Junction to Richmond and Clapham Junction - later trains will run from Stratford as far as Willesden Junction only. The following trains from Richmond to Stratford will not run: 06:08. 07:21, 08:34, 09:47, 11:01, 12:17, 13:31, 14:45, 15:56, 17:08, 18:20, 19:32, 20:47, 22:01 and 23:17
        </li>
      </ul>
    </div>
  </div>
</div>
`;
